docs/proposals/20200730-ConfigV1Alpha1.md

```diff
@@ -0,0 +1,166 @@
+#
+# The goal in this proposal is to outline the v1alpha2 config manifests and to
+# detail how multi-doc YAML enables API and config driven management of Talos
+#  without introducing accidental complexity.
+#
+# One benefit of the multi-doc approach is that we can iterate on finely scoped
+# configuration changes. If there is a fundamental design flaw in a manifest, we
+# can safely release a new version, and do so gradually.
+#
+# Over time we can start to align manifest kinds with APIs, thereby allowing us
+# users to manage Talos via the API, or by configuration. This also opens up
+# opportunities for Talos to become more dynamic. We can make decisions based
+# on the presence/absence of a manifest kind, and, in some cases, either
+# generate or retrieve from the Talos control plane.
+#
+# In this version, all manifests will be persisted locally, and never updated
+# from the `talos.config` endpoint. An API for config management will be
+# introduced that allows for updating the config. We will introduce a new
+# service, configd, that will implement the API. The API will be implemented
+# such that internal services will get their configuration using the API.
+#
+---
+# Controlplane contains configuration options for the Talos and Kubernetes
+# control planes. The absence of this manifest enables automatic retrieval from
+# the Talos control plane.
+#
+# Dynamic: true
+# Required: false
+kind: controlplane
+version: v1alpha1
+spec:
+  talos:
+    ca:
+      key: ""
+---
+# Join contains the parameters required to join the Talos and Kubernetes control
+# planes.
+#
+# Required: true
+# Dynamic: false
+kind: join
+version: v1alpha1
+spec:
+  kubernetes:
+    endpoint: https://1.2.3.4
+    token: 5dt69c.npg6duv71zwqhzbg
+    ca:
+      crt: ""
+  talos:
+    token: 5dt69c.npg6duv71zwqhzbg
+    ca: # Client needs this to validate trustd connection.
+      crt: ""
+---
+# Etcd contains configuration options for the etcd service. The presence of
+# this manifest enables the etcd service.
+#
+# Required: true
+# Dynamic: true
+kind: etcd
+version: v1alpha1
+spec: {}
+---
+# Kubelet contains configuration options for the kubelet service. In the absence
+# of this manifest, a node will pull this manifest from the Talos control plane.
+#
+# Required: True
+# Dynamic: true
+kind: kubelet
+version: v1alpha1
+spec:
+  image: docker.io/autonomy/kubelet:latest
+  args: {}
+---
+# Machine contains machine configuration options.
+#
+# Dynamic: false
+# Required: true
+kind: machine
+version: v1alpha1
+spec:
+  identity:
+    sans: []
+  bootloader:
+    cmdline: []
+  kernel:
+    parameters: {}
+  storage: {}
+  time:
+    ntp:
+      servers: {}
+  env: {}
+  registries: {}
+---
+
+---
+# Network contains network configuration options.
+#
+# Required: false
+# Dynamic: false
+kind: network
+version: v1alpha1
+spec:
+  interfaces:
+    - name: ""
+      cidrs:
+        1.2.3.4/8:
+          routes: []
+  bonds:
+    - name: ""
+      interfaces: []
```

Comment:
I wound _consider_ making `bondInterfaces` a different property from `interfaces`, since it has a different, more complicated structure which shouldn't be used or available for standard "native" interfaces.

Same idea for things like `vlanInterfaces` or `tunnelInterfaces` or what have you.

Reply:
Any proposal on what that would look like?

Reply:
I think it would be good to start with:
  - `simpleInterfaces` or just `interfaces`, since that's likely all most people will use
  - `bondInterfaces`
  - `vlanInterfaces`

~~Many (most, actually) of the fields of a `simpleInterface` would extend to `bondInterface` and `vlanInterface` such that in the backend, you may simply inherit `simpleInterface` into the fancier ones.  However, it makes sense to separate them, both from the UX side and the processing side.~~ (see below instead, but leaving this here for background thoughts)

Bond interfaces have the list of component interfaces for the bond, as well as the various parameters for the bond itself.  VLAN interfaces are similar, with a single physical interface on which they are built and a number of vlan parameters.

Down the road, there may be virtual interfaces, loopback interfaces, tunnel interfaces, etc.  There are a whole bunch of different types, which is part of the reason I think they deserve to be separate.

Rather than `simpleInterface` being the root, actually, I think it makes more sense to have an `ipInterface`, which specifies IP-related characteristics common to all network interfaces used for IP(v4+v6).  This would include things like `cidrs` and `routes` which are common to all of them.  For physical interfaces, you would extend that to include the match/selector description to define _which_ physical interface.

Another reasonably good structured network configuration system is that used by Vyatta.  It is organized in a similar way (though they break static routes into an entirely different tree, under "protocols", which I don't especially care for).